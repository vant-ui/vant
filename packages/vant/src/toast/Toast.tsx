import {
  watch,
  PropType,
  onMounted,
  onUnmounted,
  CSSProperties,
  defineComponent,
} from 'vue';

// Utils
import {
  pick,
  isDef,
  unknownProp,
  numericProp,
  makeStringProp,
  makeNumberProp,
  createNamespace,
} from '../utils';
import { lockClick } from './lock-click';

// Components
import { Icon } from '../icon';
import { Popup } from '../popup';
import { Loading, LoadingType } from '../loading';

// Types
import type { ToastType, ToastPosition } from './types';

const [name, bem] = createNamespace('toast');

const popupProps = [
  'show',
  'overlay',
  'transition',
  'overlayClass',
  'overlayStyle',
  'closeOnClickOverlay',
] as const;

export default defineComponent({
  name,

  props: {
    icon: String,
    show: Boolean,
    type: makeStringProp<ToastType>('text'),
    overlay: Boolean,
    message: numericProp,
    iconSize: numericProp,
    duration: makeNumberProp(2000),
    position: makeStringProp<ToastPosition>('middle'),
    className: unknownProp,
    iconPrefix: String,
    transition: makeStringProp('van-fade'),
    loadingType: String as PropType<LoadingType>,
    forbidClick: Boolean,
    overlayClass: unknownProp,
    overlayStyle: Object as PropType<CSSProperties>,
    closeOnClick: Boolean,
    closeOnClickOverlay: Boolean,
  },

  emits: ['update:show'],

  setup(props, { emit }) {
    let timer: NodeJS.Timeout;
    let clickable = false;

    const toggleClickable = () => {
      const newValue = props.show && props.forbidClick;
      if (clickable !== newValue) {
        clickable = newValue;
        lockClick(clickable);
      }
    };

    const updateShow = (show: boolean) => emit('update:show', show);

    const onClick = () => {
      if (props.closeOnClick) {
        updateShow(false);
      }
    };

    const clearTimer = () => clearTimeout(timer);

    const renderIcon = () => {
      const { icon, type, iconSize, iconPrefix, loadingType } = props;
      const hasIcon = icon || type === 'success' || type === 'fail';

      if (hasIcon) {
        return (
          <Icon
            name={icon || type}
            size={iconSize}
            class={bem('icon')}
            classPrefix={iconPrefix}
          />
        );
      }

      if (type === 'loading') {
        return (
          <Loading class={bem('loading')} size={iconSize} type={loadingType} />
        );
      }
    };

    const renderMessage = () => {
      const { type, message } = props;

      if (isDef(message) && message !== '') {
        return type === 'html' ? (
          <div class={bem('text')} innerHTML={String(message)} />
        ) : (
          <div class={bem('text')}>{message}</div>
        );
      }
    };

    watch(() => [props.show, props.forbidClick], toggleClickable);

    watch(
      () => [props.show, props.type, props.message, props.duration],
      () => {
        clearTimer();
        if (props.show && props.duration > 0) {
          timer = setTimeout(() => {
            updateShow(false);
          }, props.duration);
        }
      }
    );

    onMounted(toggleClickable);
    onUnmounted(toggleClickable);

    return () => (
      <Popup
        class={[
          bem([props.position, { [props.type]: !props.icon }]),
          props.className,
        ]}
        lockScroll={false}
        onClick={onClick}
        onClosed={clearTimer}
        onUpdate:show={updateShow}
        {...pick(props, popupProps)}
      >
        {renderIcon()}
        {renderMessage()}
      </Popup>
    );
  },
});
