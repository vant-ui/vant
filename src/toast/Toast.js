import { watch, onMounted, onUnmounted } from 'vue';

// Utils
import { createNamespace, isDef } from '../utils';
import { lockClick } from './lock-click';

// Components
import Icon from '../icon';
import Popup from '../popup';
import Loading from '../loading';

const [createComponent, bem] = createNamespace('toast');

export default createComponent({
  props: {
    icon: String,
    show: Boolean,
    message: [Number, String],
    duration: Number,
    className: null,
    iconPrefix: String,
    lockScroll: Boolean,
    loadingType: String,
    forbidClick: Boolean,
    overlayClass: null,
    overlayStyle: Object,
    closeOnClick: Boolean,
    type: {
      type: String,
      default: 'text',
    },
    position: {
      type: String,
      default: 'middle',
    },
    transition: {
      type: String,
      default: 'van-fade',
    },
  },

  emits: ['update:show'],

  setup(props, { emit }) {
    let timer;
    let clickable = false;

    const toggleClickable = () => {
      const newValue = props.show && props.forbidClick;
      if (clickable !== newValue) {
        clickable = newValue;
        lockClick(clickable);
      }
    };

    const onClick = () => {
      if (props.closeOnClick) {
        emit('update:show', false);
      }
    };

    const clearTimer = () => {
      clearTimeout(timer);
    };

    const renderIcon = () => {
      const { icon, type, iconPrefix, loadingType } = props;
      const hasIcon = icon || type === 'success' || type === 'fail';

      if (hasIcon) {
        return (
          <Icon
            name={icon || type}
            class={bem('icon')}
            classPrefix={iconPrefix}
          />
        );
      }

      if (type === 'loading') {
        return <Loading class={bem('loading')} type={loadingType} />;
      }
    };

    const renderMessage = () => {
      const { type, message } = props;

      if (isDef(message) && message !== '') {
        return type === 'html' ? (
          <div class={bem('text')} innerHTML={message} />
        ) : (
          <div class={bem('text')}>{message}</div>
        );
      }
    };

    watch([() => props.show, () => props.forbidClick], toggleClickable);

    watch([() => props.show, () => props.duration], () => {
      clearTimer();
      if (props.show && props.duration > 0) {
        timer = setTimeout(() => {
          emit('update:show', false);
        }, props.duration);
      }
    });

    onMounted(toggleClickable);
    onUnmounted(toggleClickable);

    return () => (
      <Popup
        show={props.show}
        class={[
          bem([props.position, { [props.type]: !props.icon }]),
          props.className,
        ]}
        lockScroll={false}
        transition={props.transition}
        overlayClass={props.overlayClass}
        overlayStyle={props.overlayStyle}
        onClick={onClick}
        onClosed={clearTimer}
      >
        {renderIcon()}
        {renderMessage()}
      </Popup>
    );
  },
});
