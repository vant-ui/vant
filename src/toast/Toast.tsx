import { watch, PropType, onMounted, onUnmounted, CSSProperties } from 'vue';

// Utils
import { createNamespace, isDef, UnknownProp } from '../utils';
import { lockClick } from './lock-click';

// Components
import Icon from '../icon';
import Popup from '../popup';
import Loading, { LoadingType } from '../loading';

const [createComponent, bem] = createNamespace('toast');

export type ToastType = 'text' | 'loading' | 'success' | 'fail' | 'html';
export type ToastPosition = 'top' | 'middle' | 'bottom';

export default createComponent({
  props: {
    icon: String,
    show: Boolean,
    overlay: Boolean,
    message: [Number, String],
    className: UnknownProp,
    iconPrefix: String,
    loadingType: String as PropType<LoadingType>,
    forbidClick: Boolean,
    overlayClass: UnknownProp,
    overlayStyle: Object as PropType<CSSProperties>,
    closeOnClick: Boolean,
    closeOnClickOverlay: Boolean,
    type: {
      type: String as PropType<ToastType>,
      default: 'text',
    },
    duration: {
      type: Number,
      default: 2000,
    },
    position: {
      type: String as PropType<ToastPosition>,
      default: 'middle',
    },
    transition: {
      type: String,
      default: 'van-fade',
    },
  },

  emits: ['update:show'],

  setup(props, { emit }) {
    let timer: NodeJS.Timeout;
    let clickable = false;

    const toggleClickable = () => {
      const newValue = props.show && props.forbidClick;
      if (clickable !== newValue) {
        clickable = newValue;
        lockClick(clickable);
      }
    };

    const updateShow = (show: boolean) => emit('update:show', show);

    const onClick = () => {
      if (props.closeOnClick) {
        updateShow(false);
      }
    };

    const clearTimer = () => {
      clearTimeout(timer);
    };

    const renderIcon = () => {
      const { icon, type, iconPrefix, loadingType } = props;
      const hasIcon = icon || type === 'success' || type === 'fail';

      if (hasIcon) {
        return (
          <Icon
            name={icon || type}
            class={bem('icon')}
            classPrefix={iconPrefix}
          />
        );
      }

      if (type === 'loading') {
        return <Loading class={bem('loading')} type={loadingType} />;
      }
    };

    const renderMessage = () => {
      const { type, message } = props;

      if (isDef(message) && message !== '') {
        return type === 'html' ? (
          <div class={bem('text')} innerHTML={String(message)} />
        ) : (
          <div class={bem('text')}>{message}</div>
        );
      }
    };

    watch(() => [props.show, props.forbidClick], toggleClickable);

    watch(
      () => [props.show, props.duration],
      () => {
        clearTimer();
        if (props.show && props.duration > 0) {
          timer = setTimeout(() => {
            updateShow(false);
          }, props.duration);
        }
      }
    );

    onMounted(toggleClickable);
    onUnmounted(toggleClickable);

    return () => (
      <Popup
        show={props.show}
        class={[
          bem([props.position, { [props.type]: !props.icon }]),
          props.className,
        ]}
        overlay={props.overlay}
        lockScroll={false}
        transition={props.transition}
        overlayClass={props.overlayClass}
        overlayStyle={props.overlayStyle}
        closeOnClickOverlay={props.closeOnClickOverlay}
        onClick={onClick}
        onClosed={clearTimer}
        {...{ 'onUpdate:show': updateShow }}
      >
        {renderIcon()}
        {renderMessage()}
      </Popup>
    );
  },
});
