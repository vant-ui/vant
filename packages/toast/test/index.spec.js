import Toast from '../';
import demoTest from '../../../test/demo-test';

demoTest(Toast);
