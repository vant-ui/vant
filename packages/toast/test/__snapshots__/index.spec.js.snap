// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Toast correctly 1`] = `
<section class="van-doc-demo-section demo-Toast" style="background:undefined;">
  <section class="van-doc-demo-block">
    <h2 class="van-doc-demo-block__title">文字提示</h2>
    <button class="van-button van-button--default van-button--normal">
      <!----><span class="van-button__text">文字提示</span></button>
  </section>
  <section class="van-doc-demo-block">
    <h2 class="van-doc-demo-block__title">加载提示</h2>
    <button class="van-button van-button--default van-button--normal">
      <!----><span class="van-button__text">加载提示</span></button>
  </section>
  <section class="van-doc-demo-block">
    <h2 class="van-doc-demo-block__title">成功/失败提示</h2>
    <button class="van-button van-button--default van-button--normal">
      <!----><span class="van-button__text">成功提示</span></button>
    <button class="van-button van-button--default van-button--normal">
      <!----><span class="van-button__text">失败提示</span></button>
  </section>
  <section class="van-doc-demo-block">
    <h2 class="van-doc-demo-block__title">高级用法</h2>
    <button class="van-button van-button--default van-button--normal">
      <!----><span class="van-button__text">高级用法</span></button>
  </section>
</section>
`;
